Login logic

stevanovicm5 opened the pull request:
- Implements basic login logic workflow with password expiration validation and JWT access token generation.
- Currently uses a hard coded secret key which should be moved to an `.env` file
- Login requires OTP code confirmation which is sent via mail service

stevanovicm5 committed:
WIP: add user login service and repo

stevanovicm5 committed:
Merge branch 'user-registration-logic' into login-logic

stevanovicm5 committed:
WIP: implement login logic

stevanovicm5 committed:
add jwt token for login logic

stevanovicm5 committed:
add password expiration validation

stevanovicm5 committed:
fix: expiration time

vukedd commented on user-service/services/user_service.go:
```diff
@@ -51,3 +62,36 @@ func (s *UserService) Register(ctx context.Context, reqDto *dtos.UserRegistratio
+	}
+
+	return user, nil
+}
```
Don't forget to handle the case when inactive users try to log in. The rest seems fine! :D

stevanovicm5 replied:
done

stevanovicm5 committed:
fix: check if user is innactive

stevanovicm5 committed:
Merge branch 'user-registration-logic' into login-logic

stevanovicm5 committed:
implement OTP mechanism

stevanovicm5 committed:
fix: expiry time in jwt

vanjmali committed:
Merge branch 'user-registration-logic' into login-logic

vanjmali commented on user-service/routers/router.go:
```diff
@@ -9,8 +9,9 @@ func HandleRequests(h *handlers.UserHandler) *mux.Router {
+	r.HandleFunc("/login", h.HandleLogin).Methods("POST")
 	// the verify endpoint is defined as a get so it can redirect when link click happens,
 	r.HandleFunc("/verify", h.HandleAccountVerification).Methods("GET", "POST")
+	r.HandleFunc("/login/verify-otp", h.HandleVerifyLoginOtp).Methods("POST")
```
For organizing purposes swap the order of handlers:


```suggestion
	r.HandleFunc("/register", h.HandleRegistration).Methods("POST")

	r.HandleFunc("/login", h.HandleLogin).Methods("POST")
	r.HandleFunc("/login/verify-otp", h.HandleVerifyLoginOtp).Methods("POST")

	// the verify endpoint is defined as a get so it can redirect when link click happens,
	r.HandleFunc("/verify", h.HandleAccountVerification).Methods("GET", "POST")
```

vanjmali commented on user-service/entities/user.go:
```diff
@@ -40,13 +40,19 @@ type User struct {
 }
 
+type OTPCode struct {
+	Content string    `bson:"content"`
```
`OTPCode.Content` can be refactored as `OTPCode.hash` to easily read what kind of data it stores.

Don't forget to also updated references in [`user_repository.go`](https://github.com/vanjmali/spotlite/blob/b216e8a8df3b8cccd04a66a5979074fe1bc1ea27/user-service/repositories/user_repository.go#L81) file to comply with the changes.

stevanovicm5 replied:
done

stevanovicm5 committed:
fix: renamed content to hash

stevanovicm5 committed:
formatting router

vanjmali reviewed (approved):
LGTM

vanjmali committed:
Add dockerized user service and local Traefik/dev stack (#7)

- Adds a re-usable multi-stage Dockerfile `Dockerfile.microservice` that
can be used for building Go services.
- Includes caching to imporve build time, especially for re-building
existing services.
- This can be used in `docker-compose.yml` file for the service in
question.
- Implements root `docker-compose.yml` and extended
`docker-compose.dev.yml` for development environment.
- Each service has its own `docker-compose.yml` that hosts the entire
service. These are all concatted togetther using `COMPOSE_FILE`
envrionment variable defined in the `.env.example` file.
- Refactored `user-service` to use Docker services and enviroment
variables
- List of used variables can be view in
`user-serivce/docker-compose.yml` file.
- All `user-service` related env variables have a `SRV_USER_` prefix to
avoid conflicts.
- Defined `api-gateway` service to use
[Traefik](https://traefik.io/traefik) as a proxy to combine all
services.
  - `frontend` will be served at the root (`/`).
- Other microservices are served at `/api/<service-name>/...` (e.g.
`/api/users/...`) that trim the prefix from the request path.
- Paths are defined via Docker container labels, see
`user-service/docker-compose.yml`
- Documented how to use, and create for new services for Docker Compose.

## Usage

1. Copy `.env.example` to `.env`
2. Populate .env with your own values if needed.
3. Start everything: `docker compose up --build`

To rebuild/restart one service (example: user-service): `docker compose
up -d --build user-service`

## Adding a new service

Each service has a `docker-compose.yml` file that gives instructions for
Docker on how to run it as well as its dependencies.

You can use the `Dockerfile.microservice` for building the service
image. See other services for examples.

The dependencies used should have a prefix, usually the name of the
service, i.e. `user-service` has a prefix `user-` (e.g. `user-mongodb`).
This avoids name conflicts with other services during development.

The `COMPOSE_FILE` variable in `.env.example` and `.env` should be
updated to include the new service, for both shells displayed.

---------

Co-authored-by: Vukasin <137070008+vukedd@users.noreply.github.com>

vukedd committed:
Merge branch 'develop' into login-logic
